src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/PosBlockSignatureRule.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using NBitcoin.Crypto;
+
+namespace Stratis.Bitcoin.Features.Consensus.Rules.CommonRules
+{
+    /// <summary>
+    /// Context checks on a POS block.
+    /// </summary>
+    public class PosBlockSignatureRule : PosConsensusRule
+    {
+        /// <inheritdoc />
+        public override Task RunAsync(RuleContext context)
+        {
+            Block block = context.BlockValidationContext.Block;
+
+            // Check proof-of-stake block signature.
+            if (!this.CheckBlockSignature(block))
+            {
+                this.Logger.LogTrace("(-)[BAD_SIGNATURE]");
+                ConsensusErrors.BadBlockSignature.Throw();
+            }
+
+
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// Checks if block signature is valid.
+        /// </summary>
+        /// <param name="block">The block.</param>
+        /// <returns><c>true</c> if the signature is valid, <c>false</c> otherwise.</returns>
+        private bool CheckBlockSignature(Block block)
+        {
+            if (BlockStake.IsProofOfWork(block))
+            {
+                bool res = block.BlockSignatur.IsEmpty();
+                this.Logger.LogTrace("(-)[POW]:{0}", res);
+                return res;
+            }
+
+            if (block.BlockSignatur.IsEmpty())
+            {
+                this.Logger.LogTrace("(-)[EMPTY]:false");
```

Comment:
All these logs are exit logs, they should not be because you don't have any entry logs, in general logging in this part of the coded cares about rules is not good

Reply:
Yeah its a good question about logging, do we log inside rules?

Reply:
We can definitely look inside rules, but but if you have an exit log should have an entry log. Also you would like to have instance logger and not to pass it from another class